kubeadm

errordeveloper opened the pull request:
**What this PR does / why we need it**:

This PR add alpha version of `kubeadm` tool, which allows user to boostrap a cluster rather quite easily. This is the initial contribution from @kubernetes/sig-cluster-lifecycle members, who's aim is to build easy-to-use tools that help to operate a cluster throughout its lifetime.

**Which issue this PR fixes**: a leap towards kubernetes/features#11

**Special notes for your reviewer**: previously seen by many folks in #31221

**Release note**:

``` release-note
`kubeadm` (alpha) provides an easy way to securely bootstrap Kubernetes on Linux, see http://kubernetes.io/docs/kubeadm/
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33262)

<!-- Reviewable:end -->

errordeveloper committed:
Initial version of kubeadm

Commit:
Fix compilation errors after method GetDefaultHostIP was removed

errordeveloper committed:
Address comments in review

- start cleaning up `cmd/manual.go`
- refine progress and error messages
- add missing blank lines after the license headers
- run `gofmt -s -w`
- do not set fake cloud provider
- add a note on why we cannot remove `HostNetwork: true` from `kube-discovery` pod just yet
- taint master and use `role=master`, set tolerations and affinity for `kube-discovery`
- parametrise log-level flag for all components

pires committed:
Added DNS add-on.

errordeveloper committed:
Refactor kube-dns addon constructors, more labels

- also add another IP to SANs
- fix mkdir calls
- add TODO for ComponentConfig
- start tagging TODOs by phases

luxas committed:
Cleanup/refactor some things, make it possible to use individual images, hide unused flags

errordeveloper committed:
Refactoring improtant parts and start on docs

Commit:
Check that sever API provides certificates capabilities

Print user friendly error with information about required version.

dgoodwin committed:
Use string bytes for token instead of decoding hex.

Commit:
Mount etcd data directory to host

errordeveloper committed:
Bump component image version tag

luxas committed:
Various improvements for kubeadm. Removed the user command, as it's too little time for implementing that. Now it's possible to use multiple arches.

Commit:
Add a flag to allow to schedule workload to the master node

errordeveloper committed:
Capture an important to-do note

Commit:
Fix Command for etcd Pod

getComponentCommand method is too hyperkube specific
to be used for etcd Pod.

errordeveloper committed:
Move `pkg/kubadm` to `cmd/kubeadm/app`, remove `cmd/manual.go`

Commit:
Add node CIDR allocation as an option to kubeadm.

This is useful for users who are used to deploying with a flannel
overlay network.

pires committed:
Reviewed help text, fix typos, go {fmt,vet,lint}.

errordeveloper committed:
Cleanup some low-hanging fruits and review TODOs

pires committed:
Add support for external and optionally secured etcd cluster.

lukemarsden committed:
Before declaring success, require that the discovery deployment has at least one active pod.

dgoodwin committed:
Allow etcd container to work with selinux.

lukemarsden commented:
I am okay with these commits being contributed to Google.

pires commented:
I am okay with these commits being contributed to Google.

sttts commented on cmd/kubeadm/README.md:
```diff
@@ -0,0 +1,97 @@
+
+- `--token=<token>`
+
+By default, a token is generated, but if you are to automate cluster deployment, you will want to
```
indent the paragraph and you get nice alignment with the flag.

sttts commented on cmd/kubeadm/README.md:
```diff
@@ -0,0 +1,97 @@
+By default, `kubeadm` sets `cluster.local` as the cluster DNS domain. If you would like to set a different one, use `--service-dns-domain`.
+
+- `--schedule-workload=<bool>` (default: "false")
+
```
strange flag name. Why not be more explicit: --master-schedulable

errordeveloper replied:
Good question, I am not entierely sure why we ended-up with this exact flag. @lukemarsden might have some background on this.

lukemarsden replied:
on balance i think `--master-schedulable` is better, i'd be happy to change this but let's be quick about it

errordeveloper replied:
Also, `-S` alias would probably make sense. To be very honest, I struggle to spell "schedulable", pronouncing it is probably even harder. On that note, I am okay to change it, but may be we can think of something a bit better? For example, `--accept-workload` may be a good one to consider.

sttts replied:
`--schedule-pods-on-master`

sttts replied:
Without the word "master" it's not clear to me that this only applies to the master nodes. We are initializing the cluster. So flags might have a cluster-wide influence.

lukemarsden replied:
Let's go with `--schedule-pods-here`.

errordeveloper replied:
I'll be looking out for having a reverse one for nodes, cause you might wanna have spare nodes that initially don't take any workload until you turn the knob. I mean: `kubeadm join --schedule-pods-here=false`.

luxas commented:
I am okay with these commits being contributed to Google.

sttts commented on cmd/kubeadm/app/cmd/cmd.go:
```diff
@@ -0,0 +1,90 @@
+
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/util/flag"
+
```
- empty line

errordeveloper replied:
Ah, there is also a bit of junk down below, I completely forgot to look at this doc, I'll go and cleanup now!

sttts commented on cmd/kubeadm/README.md:
```diff
@@ -0,0 +1,94 @@
+# Kubernetes Cluster Boostrap Made Easy
```
bootstrap

taimir commented:
I am okay with these commits being contributed to Google.

errordeveloper commented:
***Docs: https://github.com/kubernetes/kubernetes.github.io/pull/1265***

lukemarsden commented on cmd/kubeadm/README.md:
```diff
@@ -0,0 +1,94 @@
+
+By default, when `kubeadm init` runs, a token is generated and revealed in the output. That's the token you should use here.
+
+## User Experience Considerations (move to design doc, when we have one)
```
let's just delete this section

lukemarsden commented on cmd/kubeadm/app/kubeadm.go:
```diff
@@ -0,0 +1,73 @@
+		"host_pki_path":      "/etc/kubernetes/pki",
+		"host_etcd_path":     "/var/lib/etcd",
+		"hyperkube_image":    "",
+		"discovery_image":    "dgoodwin/kubediscovery:latest", // TODO(phase1): fmt.Sprintf("gcr.io/google_containers/kube-discovery-%s:%s", runtime.GOARCH, "1.0"),
```
this is blocked on https://github.com/kubernetes/kubernetes/pull/32203 and getting this image into gcr.io I think

lukemarsden replied:
@mikedanese

lukemarsden replied:
or we can leave this as-is and do a followup PR once #32203 is ready, that would be fine too IMO

luxas replied:
PR is merged, as soon as @mikedanese has pushed the images, update

mikedanese replied:
Done.

errordeveloper replied:
Could you please confirm exact image with the tag and I'll update this?

mikedanese replied:
The format in the TODO looks correct. I can confirm later.

errordeveloper replied:
Doesn't look it's present, I've tried different tags and cannot remember how to list tags. In fact, I am not even sure how to confirm `gcr.io/google_containers/kube-apiserver-amd64` exist at all.

luxas replied:
@errordeveloper https://gcr.io/v2/google-containers/kube-discovery-amd64/tags/list

I'll include it in my new patch

Thanks @mikedanese!

lukemarsden reviewed (commented):
couple of review comments

Commit:
Fix package / struct naming after core refactoring.

Commit:
Add kubeadm flags to known-flags.txt

Commit:
Rename flag `--schedule-workload` to `--schedule-pods-here` for kubeadm init

luxas committed:
Make the Kubernetes control plane version configurable

Commit:
Fix boostrap token encoding bug during master init

Currently the boostrap fails when a token is provided by the user
on `master init` and works when the token is generated. This is
because of a mismatch of how the token string in the kube-discovery
secret is encoded.

errordeveloper commented:
Verification should pass now.

errordeveloper committed:
Fix sorting of linted packages and gofmt

mikedanese commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -0,0 +1,199 @@
+
+	cmd.PersistentFlags().StringVar(
+		&s.Secrets.GivenToken, "token", "",
+		`(optional) Shared secret used to secure cluster bootstrap. If none is provided, one will be generated for you.`,
```
If all flags are optional is it useful to have (optional) in front of them?

errordeveloper replied:
Ack, will take it out.

luxas commented:
You should run `hack/update-munge-docs.sh` as well...

mikedanese commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -0,0 +1,199 @@
+}
+
+// RunInit executes master node provisioning, including certificates, needed static pod manifests, etc.
+func RunInit(out io.Writer, cmd *cobra.Command, args []string, s *kubeadmapi.KubeadmConfig, advertiseAddrs *[]string) error {
```
init.RunInit is redundant, consider init.Run

luxas replied:
This is `cmd.RunInit` and `cmd.RunJoin` also exists.
I think this is a very very small nit, not worth to prioritize

luxas replied:
Why is `advertiseAddrs` `*[]string` instead of a plain `[]string`?

errordeveloper replied:
Cobra doesn't support `[]net.IP` right now, I'm afraid, hence I've used this. Passing it by reference seemed legit, but I haven't considered the alternatives.

I wouldn't break commands into packages just yet. I can see some discussion of how `InitFlags` and `JoinFlags` can be improved, and I'd be happy to do that and reconsider how commands are organised at that stage, which would quite appropriate to do in a follow-up PR.

luxas replied:
Can't you just pass them by value for the time being (i.e. `[]string`)?

errordeveloper replied:
I don't think there is a chance for it to become `nil`, which sounds like the only real concern. I've added a TODO comment to eliminate the need for it by using custom flag parsing helper (or just make a PR to `spf13/pflag`).

mikedanese commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -0,0 +1,199 @@
+	if s.InitFlags.CloudProvider != "" {
+		// TODO(phase2) we should be able to auto-detect it and check whether things like IAM roles are correct
+		if _, ok := kubeadmapi.SupportedCloudProviders[s.InitFlags.CloudProvider]; !ok {
+			return fmt.Errorf("<cmd/init> cloud provider %q is not supported, you can use any of %v, or leave it unset", s.InitFlags.CloudProvider, kubeadmapi.ListOfCloudProviders)
```
Why are we adding `<cmd/init>` to these?  why aren't we adding `<cmd/init>` to the errors below?

errordeveloper replied:
The errors below have prefix from the package they originate from. This one originates here, it's purely about args.

errordeveloper replied:
Weird, Github doesn't actually render these strings for me: https://www.dropbox.com/s/8n8ila79924i38a/Screenshot%202016-09-23%2020.21.02.png?dl=0

errordeveloper replied:
Ok, I don't know how to do this. I'll add a note and make sure to address it in a follow-up PR.

mikedanese commented on cmd/kubeadm/app/master/discovery.go:
```diff
@@ -0,0 +1,136 @@
+	return kd
+}
+
+func CreateDiscoveryDeploymentAndSecret(s *kubeadmapi.KubeadmConfig, client *clientset.Clientset, caCert *x509.Certificate) error {
```
Why isn't this a static pod? It seems like it would make the transition into core apiserver functionality   more straightforward.

luxas replied:
I'm also in favor for a static pod here, we should just run it alongside the apiserver, cm and scheduler

errordeveloper replied:
> Why isn't this a static pod?

It uses a secret, which depends on control plane, it doesn't make any
sense to split these two things. However, the main principle driving this
decision was to mod away from static pods where possible. Not using secret is also bad, as we have to write more files to disk, and I believe it's something we prefer to avoid (especially for new pieces). We can change how `kube-discovery` is configured, e.g. be use flags, but that'd make token rotation problem look different too.

> It seems like it would make the transition into core apiserver functionality more straightforward.

The status of whether it may become part the apiserver is still uncertain.

krousey replied:
Correct me if I'm wrong, but can't you create pods that reference secrets that aren't defined? And when they become defined later, they can resolve them? So this could be a static pod that just wouldn't work correctly when the control plane first came up, but would eventually get its secret.

errordeveloper replied:
Ok, surely you can do that. It just doesn't make sense to me to separate these two things. I really don't see a point of having this discussion right now. I'd much rather discuss the future of the JWS service, i.e. whether it could be an API  object behind a feature gate or what not. In my understanding, if does become part of the API, we will just remove this code, not need to keep reshaping it. Also, I really don't understand why to use a static pod after we have established the control plane, removing the node affinity and tolerance annotation would be the only thing, but why would we not dogfood those features?

Additionally, making this a static pod reduces upgradability, as it progresses to stable we might have to make it easy to upgrade or simply delete (if user doesn't want to run unstable components after initial cluster was built).

mikedanese commented on pkg/storage/etcd/etcdtest/doc.go:
```diff
@@ -1,17 +0,0 @@
-limitations under the License.
-*/
-
-package etcdtest // import "k8s.io/kubernetes/pkg/storage/etcd/etcdtest"
```
why?

errordeveloper replied:
Strange rebase artefact, will have to check.

errordeveloper replied:
I've fix this.

mikedanese commented:
I haven't looked at all of this yet.

mikedanese commented on cmd/kubeadm/README.md:
```diff
@@ -0,0 +1,63 @@
+# Kubernetes Cluster Bootstrap Made Easy
```
This doc should be in the docs repo. None of the other commands have docs here. We are trying to consolidate where we put user docs.

errordeveloper replied:
Ack, will move it.

errordeveloper replied:
Removed.

errordeveloper commented:
Mike, thanks for your comments!

We are using taints. The flag is handy, it is not very simple to unset the
taints.

I shall let @pires address external etcd stuff, he has insisted to add
support for it.

On Fri, 23 Sep 2016, 19:33 Mike Danese, <notifications@github.com> wrote:

> I haven't looked at all of this yet.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/33262#issuecomment-249269098>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPWS2lzBYLKWFkXBvqlKtHI5_vFz1jdks5qtBuBgaJpZM4KDq18>
> .
>

mikedanese commented:
I don't think we ever want to unset the taint?

errordeveloper commented:
We have heard several requests to allow for single-node use cases, one of
the users mentioned that minikube doesn't cut it for them.

On Fri, 23 Sep 2016, 19:41 Mike Danese, <notifications@github.com> wrote:

> I don't think we ever want to unset the taint?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/33262#issuecomment-249271018>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPWS-Q-Ly35mm2Ag7QRIdi3nUOgaoazks5qtB1OgaJpZM4KDq18>
> .
>

mikedanese commented:
@krousey @roberthbailey

errordeveloper commented:
It needs a secret, which depends on control plane, it doesn't make any
sense to split these two things. However, the main principle driving this
decision was to mod away from static pods where possible.

On Fri, 23 Sep 2016, 20:07 Lucas Käldström, <notifications@github.com>
wrote:

> *@luxas* commented on this pull request.
> ------------------------------
>
> In cmd/kubeadm/app/master/discovery.go
> <https://github.com/kubernetes/kubernetes/pull/33262>:
>
> > +	kd := kubeDiscovery{
> +		Deployment: NewDeployment(kubeDiscoveryName, 1, newKubeDiscoveryPodSpec(s)),
> +		Secret: &api.Secret{
> +			ObjectMeta: api.ObjectMeta{Name: kubeDiscoverySecretName},
> +			Type:       api.SecretTypeOpaque,
> +			Data:       encodeKubeDiscoverySecretData(s, caCert),
> +		},
> +	}
> +
> +	SetMasterTaintTolerations(&kd.Deployment.Spec.Template.ObjectMeta)
> +	SetMasterNodeAffinity(&kd.Deployment.Spec.Template.ObjectMeta)
> +
> +	return kd
> +}
> +
> +func CreateDiscoveryDeploymentAndSecret(s *kubeadmapi.KubeadmConfig, client *clientset.Clientset, caCert *x509.Certificate) error {
>
> I'm also in favor for a static pod here, we should just run it alongside
> the apiserver, cm and scheduler
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/33262>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPWS9aO6CebS5a-KKc2gyTwczUbC-nNks5qtCOKgaJpZM4KDq18>
> .
>

luxas commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -0,0 +1,199 @@
+	)
+
+	// TODO (phase1+) @errordeveloper make the flags below not show up in --help but rather on --advanced-help
+	cmd.PersistentFlags().StringSliceVar(
```
I was thinking about if we could make it only two flags, `--external-etcd-endpoints` and `--external-etcd-certdir` with hardcoded cert/keyfile/ca filenames

errordeveloper replied:
Again, I'll let @pires comment on this, as he insisted to add this feature.

errordeveloper replied:
I am all in for trimming flags. I'd start by grouping `--help|--basic-usage` and `--advanced-usage`, and provide config file for anything beyond basic eventually.

luxas commented on cmd/kubeadm/app/api/types.go:
```diff
@@ -0,0 +1,101 @@
+
+// InitFlags holds values for "kubeadm init" command flags.
+type InitFlags struct {
+	API struct {
```
Is that a blocker for Monday/Tuesday or can we do that later?

luxas commented on cmd/kubeadm/app/api/types.go:
```diff
@@ -0,0 +1,101 @@
+
+const (
+	DefaultServiceDNSDomain   = "cluster.local"
+	DefaultServicesCIDRString = "100.64.0.0/12"
```
I've voted and am still voting for `10.16.0.0/16`

errordeveloper replied:
See https://github.com/kubernetes/kubernetes/pull/33262/files#r80358971 for what it is and why. It's hard to pick this, the idea is that we pick the safer one for services and leave pod network to decide what should be used for pods.  I think this Carrier-grade NAT CIDR fits the bill very well for services. If we think it's polically incorrect, that would a different matter. So far I haven't heard an argument agains using this range other then those that arise from confusion with non-private ranges, it's indeed a kind of range we can use (see links in my other comment).

errordeveloper replied:
@jbeda I think we could indeed make it more explicit in the output, also pre-flight checker could do some lightweight network intelligence for us... Also docs should highlight this clearly, of course, along with reasoning to what the defaults are.

mikedanese replied:
My understanding is that the range is for use by ISPs, not for private use. Are we already picked a private range for pods? Can we just use the next /16?

errordeveloper replied:
> Aren't we already picked a private range for pods?

No, actually we leave that up to the user, as we only need it for flannel. Weave certainly want pod network to not be defined, and, IIRC, Calico too.

mikedanese replied:
Both AWS and Google networks use this range. This is going to cause weird failures.

errordeveloper commented:
ping @jbeda

luxas commented on cmd/kubeadm/app/api/types.go:
```diff
@@ -0,0 +1,101 @@
+		DNSDomain string
+	}
+	PodNetwork struct {
+		CIDR net.IPNet
```
Moving everything into the same `Network` struct is fine

luxas commented on cmd/kubeadm/app/api/types.go:
```diff
@@ -0,0 +1,105 @@
+
+func init() {
+	_, DefaultServicesCIDR, _ = net.ParseCIDR(DefaultServicesCIDRString)
+	SupportedCloudProviders = make(map[string]bool, len(ListOfCloudProviders))
```
I think this is unnecessary. You should just check if the given cloudprovider is existent in `ListOfCloudProviders` in a Go-native way

errordeveloper replied:
I've been told that `map[string]bool` is Go-native. What are you suggesting exactly?

luxas replied:
That
a) You're looping through `ListOfCloudProviders` when validating the flag or
b) Making ListOfCloudProviders a `map[string]bool` from the start in order to be able to directly check if the cloudprovider exists

luxas commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -0,0 +1,199 @@
+	advertiseAddrs := &[]string{}
+	cmd := &cobra.Command{
+		Use:   "init",
+		Short: "Run this on the first machine.",
```
`Run this in order to set up the Kubernetes master`

errordeveloper replied:
Ack.

luxas commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -0,0 +1,199 @@
+	)
+	cmd.PersistentFlags().StringVar(
+		&s.InitFlags.CloudProvider, "cloud-provider", "",
+		`(optional) Enable a specific cloud provider features (external load-balancers, storage, etc), e.g. "gce"`,
```
remove `a specific`

errordeveloper replied:
Ack.

luxas commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -0,0 +1,199 @@
+	)
+	cmd.PersistentFlags().StringVar(
+		&s.InitFlags.API.Etcd.ExternalCAFile, "external-etcd-cafile", "",
+		`(optional) etcd certificate authority certificate file."`,
```
nit: ExternalCAFile, ExternalCertFile, ExternalKeyFile descriptions end with `"`

errordeveloper replied:
Ack.

luxas commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -0,0 +1,199 @@
+		return err
+	}
+
+	kubeconfigs, err := kubemaster.CreateCertsAndConfigForClients(s, []string{"kubelet", "admin"}, caKey, caCert)
```
Why is both `kubelet.json` and `admin.json` created? How do one differ from the other?

errordeveloper replied:
@jbeda yes, admin config is sort of recovery path, user may wish to copy it to a safe place etc, we can address it in the docs. Also, `kubeadm user` is still on the table, it'd do CSR dance for users etc. May be we could consider Google apps auth also, but in any case admin/root key seems like a handy thing to have, given we are still in alpha and who know what may have go wrong with CSR API.

luxas replied:
Okay, so `admin.json` is unused for the time being?
IIRC they're equal at the moment as well

So we're leaving it for users and will document that for now the user should copy `/etc/kubernetes/admin.json` and use it for authenticating with the apiserver?

That SGTM

errordeveloper replied:
> IIRC they're equal at the moment as well

No, the files contain different client certs.

> So we're leaving it for users and will document that for now the user should copy /etc/kubernetes/admin.json and use it for authenticating with the apiserver?

Yes.

luxas commented on cmd/kubeadm/app/master/manifests.go:
```diff
@@ -0,0 +1,302 @@
+		Name: "certs",
+		VolumeSource: api.VolumeSource{
+			// TODO(phase1+) make path configurable
+			HostPath: &api.HostPathVolumeSource{Path: "/etc/ssl/certs"},
```
@pires We should definitely document that the `--external-etcd-cafile`, `--external-etcd-certfile` and `--external-etcd-keyfile` **only can** start with `/etc/ssl/certs` or just dump the individual parameters in favor for `--external-etcd-certdir` and fix this

luxas commented on cmd/kubeadm/app/master/pki.go:
```diff
@@ -0,0 +1,181 @@
+func newServerKeyAndCert(s *kubeadmapi.KubeadmConfig, caCert *x509.Certificate, caKey *rsa.PrivateKey, altNames certutil.AltNames) (*rsa.PrivateKey, *x509.Certificate, error) {
+	key, err := certutil.NewPrivateKey()
+	if err != nil {
+		return nil, nil, fmt.Errorf("unabel to create private key [%s]", err)
```
`unable to create private key [%v]`

jbeda replied:
Make sure you get the other occurrences of this in this file.

errordeveloper replied:
Thanks, I will. Didn't think it was very critical to do, but as you insist...

ekozan commented:
single-node is needed 👍

luxas commented on cmd/kubeadm/app/master/pki.go:
```diff
@@ -0,0 +1,181 @@
+
+	internalAPIServerVirtualIP, err := ipallocator.GetIndexedIP(&s.InitFlags.Services.CIDR, 1)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to allocate IP address for the API server from the given CIDR (%q) [%s]", &s.InitFlags.Services.CIDR, err)
```
every error here should start with `<master/pki>` and use `%v` for error formatting

errordeveloper replied:
I've made all `fmt.Errorf()`'s use `%v` now. Please take a closer look, you will see `<master/pki>` is added in the final `fmt.Errorf()`.

luxas commented on cmd/kubeadm/app/master/manifests.go:
```diff
@@ -0,0 +1,302 @@
+			"--insecure-experimental-approve-all-kubelet-csrs-for-group=system:kubelet-bootstrap",
+		},
+		scheduler: {
+			// TODO(phase1+): consider adding --address=127.0.0.1 in order to not expose the scheduler port to the rest of the world
```
I really think this should be fixed. It would be strange to advertise we're setting up the cluster securely while exposing controller-manager and scheduler to the rest of the world if the apiserver is accessible to the outside

errordeveloper replied:
I assume you don't really mean "rest of the world" in the most literal sense... Didn't you change this in your PR (which I've reviewed and am waiting for your response)?

mikedanese replied:
controller-manager and scheduler rest handlers expose basically nothing.

mikedanese replied:
We also shouldn't advertise that we are setting up the cluster securely.

lukemarsden replied:
I'll update the docs to caveat the secure claim with a reiteration of the fact that it's in alpha.

ekozan commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -0,0 +1,196 @@
+		`(optional) Enable a specific cloud provider features (external load-balancers, storage, etc), e.g. "gce"`,
+	)
+	cmd.PersistentFlags().BoolVar(
+		&s.InitFlags.Schedulable, "schedule-pods-here", false,
```
Please don't remove single node possibility

mikedanese replied:
@ekozan how does adding a taint to the master remove single node possibility? I don't know why you would ever not want a taint to mark the master, just tolerate it in your workload. If you reaallly hate it, you  can remove it with a single kubectl command.
```
$ kubectl taint master-node -dedicated
```

errordeveloper replied:
I wasn't aware of `kubectl taint` being a thing. I'd be happy to drop the flag and document this. // cc @lukemarsden

errordeveloper replied:
https://github.com/kubernetes/kubernetes/pull/33262#issuecomment-249360611

mikedanese replied:
> Re kubectl taint: I think it's ok to keep this as an explicit part of the
> UI even if there's another way to do it.

@lukemarsden If it's not clear, I'm aggressively trying to minimize the options surface for kubeadm. This response seems ad-hoc'ish to me and is not satisfying. What types of command line options are we going to choose and why? Why are we going to include this one? Why aren't we exposing --allow-privileged or --authroization-mode?

mikedanese replied:
> @lukemarsden @mikedanese I am thinking that the main benefit of having the flag for making master schedulable is about how quickly user learns what they need to do, i.e. they don't have to read the docs.

This is still ad-hoc and there is no clear way of extending this thinking to other cases in the future. I can apply the same logic to *many* other options. Is that what we want?

mikedanese replied:
If anything it seems more useful to expose options that can't be changed through kubectl.

errordeveloper replied:
@mikedanese I largely agree with you, was just trying to find a logical UX-minded justification to doing otherwise. I'm happy to remove this flag, knowing that it's only one `kubeclt` call away (I thought `kubectl taint` didn't make it into 1.4 and one had to do `kubectl edit nodes` or something).

mikedanese replied:
I'm just trying to establish a framework for how we decide whether to include/exclude options int the future. If there is already a framework, it's not clear to me. I'm definitely not arguing for including those other flags or even necessarily excluding this one. I'm making a point to discuss this now because what we decide will be a precedent for later changes. Do you a agree that it would be a good idea to have a clear guideline for what we want to include in the set of configurable options?

lukemarsden replied:
Re-posting in context:

@mikedanese my framework for what belongs in flags right now is "as little
as possible to support the Phase I requirements" -- I admit those reqs seem
to have expanded slightly to include "supports single node dev clusters
with scheduleable masters" but I'm OK with that :)

@errordeveloper if you can remove the flag, I'll document `kubectl taint master-node -dedicated`.

jbeda replied:
I don't think it is absolutely critical, but I think that we should maintain this flag.

Users will understand the concept of "schedule workloads on my master node" vs. "don't schedule normal workloads on my master node".  Users starting out *won't* necessarily understand what a taint is.  Even though it is one command away, that is an extra command that you'll have to run in the "single node" case.  And while it is also true that users can schedule general work here by having the workload tolerate, that'll be some magic voodoo to put in an already too verbose manifest. 

I'd love to avoid introducing advanced concepts like taints/tolerations when users are just starting out.

luxas reviewed (commented):
minikube has settled on `8443` as the apiserver port. 
Maybe we should do the same. `443` is bad because it's very generic.

`6443` is also fine for me

krousey commented on cmd/kubeadm/app/master/tokens.go:
```diff
@@ -0,0 +1,62 @@
+func CreateTokenAuthFile(s *kubeadmapi.KubeadmConfig) error {
+	tokenAuthFilePath := path.Join(s.EnvParams["host_pki_path"], "tokens.csv")
+	if err := generateTokenIfNeeded(s); err != nil {
+		return fmt.Errorf("<master/tokens> failed to generate token(s) [%s]", err)
```
This message could be misleading if the given token fails validation. In that cause we would indicate that we failed to generate a token when we didn't even attempt it.

errordeveloper replied:
I'm not entierly happy with this as it stand in many ways. We haven't yet decided on this, I'm hoping it will change how we handle this all together very soon.

krousey commented on cmd/kubeadm/app/master/tokens.go:
```diff
@@ -0,0 +1,62 @@
+)
+
+func generateTokenIfNeeded(s *kubeadmapi.KubeadmConfig) error {
+	ok, err := kubeadmutil.UseGivenTokenIfValid(s)
```
I know it won't happen with the way it is currently implemented, but this doesn't handle case where `ok` is true and `err` is non-nil.

errordeveloper replied:
Will add a comment, I didn't know this.

krousey commented on cmd/kubeadm/app/master/tokens.go:
```diff
@@ -0,0 +1,62 @@
+	return nil
+}
+
+func CreateTokenAuthFile(s *kubeadmapi.KubeadmConfig) error {
```
Perhaps a better overall flow would be

```go
func CreateTokenAuthFile(s *kubeadmapi.KubeadmConfig) error {
	if s.Secrets.GivenToken == "" {
		s.Secrets = kubeadmutil.GenerateToken()
	} else {
		s.Secrets = kubeadmutil.ParseToken(s.Secrets.GivenToken)
	}
	// the rest of the file handling
}
```

(error handling omitted)

Then you wouldn't need `generateTokenIfNeeded`.

jbeda replied:
Also, document public function.

errordeveloper replied:
May be, but I wouldn't risk this in the last moment. I'll add a doc string, although docs are missing in many other places, so I'm not quite sure it helps us to spend time on this right now.

krousey commented on cmd/kubeadm/app/master/manifests.go:
```diff
@@ -0,0 +1,302 @@
+	}
+}
+
+func getComponentCommand(component string, s *kubeadmapi.KubeadmConfig) (command []string) {
```
I would split this up into 3 separate commands, one for each component. Other than the hyperkube logic, they really don't share much, and should probably be separate.

krousey commented on cmd/kubeadm/app/master/pki.go:
```diff
@@ -0,0 +1,181 @@
+		altNames certutil.AltNames
+	)
+
+	if len(s.InitFlags.API.AdvertiseAddrs) > 0 {
```
You don't need to check this. `append` will just do the RightThing™

errordeveloper replied:
Ack.

krousey commented on cmd/kubeadm/app/master/pki.go:
```diff
@@ -0,0 +1,181 @@
+	return key, nil
+}
+
+func CreatePKIAssets(s *kubeadmapi.KubeadmConfig) (*rsa.PrivateKey, *x509.Certificate, error) {
```
Can you document this?

errordeveloper replied:
Will do, it's big enough.

krousey commented on cmd/kubeadm/app/master/apiclient.go:
```diff
@@ -0,0 +1,220 @@
+		n.ObjectMeta.Annotations[api.TaintsAnnotationKey] = string(taintsAnnotation)
+	}
+
+	if _, err := client.Nodes().Update(n); err != nil {
```
Change this to a loop instead of recursion.

errordeveloper replied:
Ack.

krousey reviewed (commented):
I made it through about half of the code. Sorry if some of my comments seem nit-picky.

jbeda commented on cmd/kubeadm/app/master/manifests.go:
```diff
@@ -0,0 +1,302 @@
+)
+
+// Static pod definitions in golang form are included below so that `kubeadm
+// init master` and `kubeadm manual bootstrap master` can get going.
```
nit: `kubeadm manual` is no longer around, yes?

errordeveloper replied:
Yes, it's gone and will probably re-surface as a few extra/advanced flags in `kubeadm join`. Only flag to add to `kubeadm init` would be about disabling JWS (maybe).

errordeveloper replied:
Will update the comment.

jbeda commented on cmd/kubeadm/app/master/pki.go:
```diff
@@ -0,0 +1,181 @@
+	}
+	cert, err := certutil.NewSignedCert(config, key, caCert, caKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to sing certificate [%s]", err)
```
s/sing/sign/

errordeveloper replied:
Fixed.

jbeda commented on cmd/kubeadm/app/master/tokens.go:
```diff
@@ -0,0 +1,62 @@
+		return fmt.Errorf("<master/tokens> failed to create directory %q [%s]", s.EnvParams["host_pki_path"], err)
+	}
+	serialized := []byte(fmt.Sprintf("%s,kubeadm-node-csr,%s,system:kubelet-bootstrap\n", s.Secrets.BearerToken, uuid.NewUUID()))
+	if err := cmdutil.DumpReaderToFile(bytes.NewReader(serialized), tokenAuthFilePath); err != nil {
```
Might be worth commenting that `DumpReaderToFile` will open the file as `0600`.  (And that function should probably have that in the name or take a parameter or at least document the behavior)

errordeveloper replied:
Thanks, good point.

jbeda commented on cmd/kubeadm/app/node/csr.go:
```diff
@@ -0,0 +1,117 @@
+	}
+	csrClient := client.CertificateSigningRequests()
+
+	// TODO(phase1+) checkCertsAPI() has a side-effect of making first attempt of communicating with the API,
```
This is going to be super critical.  It might be worth early in the `kube join` path to have an explicit master connectivity check so that we can handle that gracefully.  At least get to the point where the master gives us a 403.

errordeveloper replied:
Indeed, hence the todo note 👍

jbeda commented on cmd/kubeadm/app/node/csr.go:
```diff
@@ -0,0 +1,117 @@
+
+	key, err := certutil.MakeEllipticPrivateKeyPEM()
+	if err != nil {
+		return nil, fmt.Errorf("<node/csr> failed to generating private key [%s]", err)
```
Use `%v` for errors.

errordeveloper replied:
Ack. Will replace those.

jbeda reviewed (commented):
This all looks really good to me.  I have no major changes to request and so won't gate the submit.  Good work!

errordeveloper commented:
@jbeda @krousey thanks for reading through this and all the comments and suggestions. I'll reply to each and address the most trivial ones today. Hopefully we can figure out what's most important to get this merged and official alpha binary to be published.

errordeveloper commented on cmd/kubeadm/app/master/addons.go:
```diff
@@ -0,0 +1,267 @@
+	fmt.Println("<master/addons> created essential addon: kube-proxy")
+
+	kubeDNSDeployment := NewDeployment("kube-dns", 1, createKubeDNSPodSpec(s))
+	SetMasterTaintTolerations(&kubeDNSDeployment.Spec.Template.ObjectMeta)
```
It's not affinity, just toleration, i.e. it may if it has to. It seems like DNS is critical enough.

jbeda replied:
For efficiency sake, it might be best to pin it to the master.  If it gets scheduled on another node instead of the master then it is taking up space that could be used by a workload.

errordeveloper replied:
Ah, you are actually suggesting to add master affinity, that makes sense, but do we need this right now?

mikedanese replied:
@jbeda I'd prefer to leave it as is as running kube-dns on the cluster is the standard deployment. I don't see a good reason to diverge from that.

errordeveloper replied:
@mikedanese I agree.

> If it gets scheduled on another node instead of the master then it is taking up space that could be used by a workload.

This sounds like all the add-ons would have to run on master...

errordeveloper commented on cmd/kubeadm/app/master/addons.go:
```diff
@@ -0,0 +1,267 @@
+func CreateEssentialAddons(s *kubeadmapi.KubeadmConfig, client *clientset.Clientset) error {
+	arches := [3]string{"amd64", "arm", "arm64"}
+
+	for _, arch := range arches {
```
@luxas added this, I let him comment. I'd be happy if we defaulted to current arch. We also have a solution for using new multi-arch image capability, but it only works with Docker Hub.

luxas replied:
This will be here and will be documented as well.
It's here until gcr.io supports v2 schema 2 (cc @aronchick): https://twitter.com/aronchick/status/779048421738946560

Also read: https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/multi-platform.md which is accepted if you want more details.

I totally agree we should end up with manifest lists, but that's blocked for now.

kubeadm will/should be a opinionated way to set up Kubernetes on the other architectures as well, and there are probably more folks using Kubernetes on arm/arm64/ppc64le than you think

mikedanese replied:
I vote for using current architecture. That seems most reasonable until gcr supports v2 schema 2.

errordeveloper replied:
> I think creating 3 daemonsets will confuse users.

I've had exactly the same concern, and I've raised it with @luxas already.

@luxas could you please make a PR ASAP to update this, once it's not just me thinking this?

errordeveloper commented on cmd/kubeadm/README.md:
```diff
@@ -0,0 +1,63 @@
+  to detect, you can use `--api-advertise-addresses` and `--api-external-dns-names` to add multiple
+  different IP addresses and hostnames (DNS).
+
+- `--service-cidr=<cidr>` (default: "100.64.0.0/12")
```
It's not a public CIDR, it's so-called ["carrier-grade NAT"](https://en.wikipedia.org/wiki/Carrier-grade_NAT), by using this we have less chance to collide with existing private networks in use, and more options to pick for pod network. Also, we can grab a `/12` quite safely, which is the maximum API server supports (for whatever reason that is).

errordeveloper commented on cmd/kubeadm/app/cmd/cmd.go:
```diff
@@ -0,0 +1,89 @@
+	// TODO(phase2) detect interactive vs non-interactive use and adjust output accordingly
+	// i.e. make it automation friendly
+	//
+	// TODO(phase2) create an abstraction that defines files and the content that needs to
```
I believe this use-case is captured down below.

errordeveloper commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -0,0 +1,199 @@
+	)
+	cmd.PersistentFlags().StringSliceVar(
+		advertiseAddrs, "api-advertise-addresses", []string{},
+		`(optional) The IP addresses to advertise, in case autodetection fails.`,
```
Ack. That is indeed much better.

lukemarsden commented:
Re kubectl taint: I think it's ok to keep this as an explicit part of the
UI even if there's another way to do it.

On Saturday, September 24, 2016, Ilya Dmitrichenko <notifications@github.com>
wrote:

> *@errordeveloper* commented on this pull request.
> ------------------------------
>
> In cmd/kubeadm/app/cmd/init.go
> <https://github.com/kubernetes/kubernetes/pull/33262>:
>
> > +			kubeadmapi.DefaultServicesCIDRString,
> +	)
> +	cmd.PersistentFlags().IPNetVar(
> +		&s.InitFlags.PodNetwork.CIDR, "pod-network-cidr", net.IPNet{},
> +		`(optional) Specify range of IP addresses for the pod network. If set, the control plane will automatically allocate CIDRs for every node.`,
> +	)
> +	cmd.PersistentFlags().StringVar(
> +		&s.InitFlags.Services.DNSDomain, "service-dns-domain", kubeadmapi.DefaultServiceDNSDomain,
> +		`(optional) Use alternative domain for services, e.g. "myorg.internal"`,
> +	)
> +	cmd.PersistentFlags().StringVar(
> +		&s.InitFlags.CloudProvider, "cloud-provider", "",
> +		`(optional) Enable a specific cloud provider features (external load-balancers, storage, etc), e.g. "gce"`,
> +	)
> +	cmd.PersistentFlags().BoolVar(
> +		&s.InitFlags.Schedulable, "schedule-pods-here", false,
>
> I wasn't aware of kubectl taint being a thing. I'd be happy to drop the
> flag and document this. // cc @lukemarsden
> <https://github.com/lukemarsden>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/33262>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAQJ0qlkz8Et8lwlyQhL1V4BZrIzW2zhks5qtQwBgaJpZM4KDq18>
> .
>


-- 
lukemarsden.net
me@lukemarsden.net
@lmarsden <https://twitter.com/lmarsden>

errordeveloper commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -0,0 +1,199 @@
+		for _, i := range *advertiseAddrs {
+			addr := net.ParseIP(i)
+			if addr == nil {
+				return fmt.Errorf("<cmd/init> failed to parse flag (%q) as an IP address", "--api-advertise-addresses="+i)
```
Acks, and thanks. When I wrote this, I didn't realise the flag parser handles comma-separated values, I've thought user would have to pass each flag multiple times to get there.

errordeveloper commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -0,0 +1,199 @@
+	}
+
+	if s.InitFlags.CloudProvider != "" {
+		// TODO(phase2) we should be able to auto-detect it and check whether things like IAM roles are correct
```
This was note prior to our discussion which led to https://github.com/kubernetes/features/issues/88. I think add-ons are the way forward, and we shouldn't bother with auto-detection. I'll delete the note.

errordeveloper commented:
@lukemarsden email comments are detached from the thread. I'll create a pointer.

errordeveloper committed:
Address comments in review

errordeveloper commented:
I've addressed most I could today. Waiting on @pires and @luxas on response regarding external etcd and multi-arch features respectively.

errordeveloper commented:
Can we priorities what must be fixed for this to get merged, or decide whether we will need to go the alternative route, i.e. publish packages prior to merging this, and make sure everyone is happy with the code before it goes into master later next week.

errordeveloper commented:
@lukemarsden @mikedanese I am thinking that the main benefit of having the flag for making master schedulable is about how quickly user learns what they need to do, i.e. they don't have to read the docs.

errordeveloper commented on cmd/kubeadm/app/api/types.go:
```diff
@@ -0,0 +1,105 @@
+// KubeadmConfig TODO add description
+type KubeadmConfig struct {
+	InitFlags
+	JoinFlags
```
I get this, hope TODO will suffice for now?

errordeveloper commented on cmd/kubeadm/app/api/types.go:
```diff
@@ -0,0 +1,101 @@
+			ExternalKeyFile   string
+		}
+	}
+	Services struct {
```
Can I turn this into a TODO and follow-up with PR to refactor it later on?

lukemarsden commented:
@mikedanese my framework for what belongs in flags right now is "as little
as possible to support the Phase I requirements" -- I admit those reqs seem
to have expanded slightly to include "supports single node dev clusters
with scheduleable masters" but I'm OK with that :)

If you can tell me what the exact 'kubectl taint' command to get the same
behaviour as --schedule-pods-here is I'm happy for us to kill the flag and
document the taint command on the basis of "as little as possible".

Thanks for all your hard work on this!

Do you think we'll be able to get kubeadm into the 1.4 release @mikedanese?

On Sat, 24 Sep 2016, 19:28 Mike Danese, <notifications@github.com> wrote:

> *@mikedanese* commented on this pull request.
> ------------------------------
>
> In cmd/kubeadm/app/kubeadm.go
> <https://github.com/kubernetes/kubernetes/pull/33262>:
>
> > +	kubeadm: please create issue an using https://github.com/kubernetes/kubernetes/issues/new
> +	kubeadm: and make sure to mention @kubernetes/sig-cluster-lifecycle. Thank you!
> +`)
> +
> +// TODO(phase2) use componentconfig
> +// we need some params for testing etc, let's keep these hidden for now
> +func getEnvParams() map[string]string {
> +
> +	envParams := map[string]string{
> +		// TODO(phase1+): Mode prefix and host_pki_path to another place as constants, and use them everywhere
> +		// Right now they're used here and there, but not consequently
> +		"kubernetes_dir":     "/etc/kubernetes",
> +		"host_pki_path":      "/etc/kubernetes/pki",
> +		"host_etcd_path":     "/var/lib/etcd",
> +		"hyperkube_image":    "",
> +		"discovery_image":    "dgoodwin/kubediscovery:latest", // TODO(phase1): fmt.Sprintf("gcr.io/google_containers/kube-discovery-%s:%s", runtime.GOARCH, "1.0"),
>
> The format in the TODO looks correct. I can confirm later.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/33262>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAQJ0prgtYAzaR78DjZAlCCJkyDZcB9Gks5qtWu6gaJpZM4KDq18>
> .
>

errordeveloper commented:
@lukemarsden see https://github.com/kubernetes/kubernetes/pull/33262#discussion_r80345498

errordeveloper commented on cmd/kubeadm/app/util/tokens.go:
```diff
@@ -0,0 +1,85 @@
+	return nil
+}
+
+func UseGivenTokenIfValid(s *kubeadmapi.KubeadmConfig) (bool, error) {
```
This makes a lot of sense to me, and the way flags struct is passed around is a very poor design, which is begging for improvement. I believe I've captured all your comments as TODOs now.

lukemarsden commented:
> Can we prioritise what must be fixed for this to get merged?

I propose that we address only minor review comments, and defer anything larger to follow-up issues. We should do this so that we can get `kubeadm` merged into master, get packages built, tested and released as part of 1.4 on Monday. How does that sound to you @mikedanese?

BTW I have some rough [integration tests](https://github.com/lukemarsden/kubeadm-integration-tests) which I will run as soon as we have packages to give us confidence that they are working.

luxas commented:
@lukemarsden Still we should get in https://github.com/errordeveloper/kubernetes/pull/17

And refactor the individual etcd cert args into cert dir, then I think we are in good shape

simonswine commented:
Thanks for this amazing work this really makes things way simpler.

I was just wondering if the kubernetes release binaries would contain kubeadm. This would be enormously helpful to get more people involved. (At least a build.sh make did not built the binary)

I have just spun up a cluster on CoreOS and it worked like a charm. Will add a couple of instructions for the docs PR

lukemarsden commented:
I have left some comments on https://github.com/errordeveloper/kubernetes/pull/17, basically I think we should lgtm this as-is (after removing `--schedule-pods-here`) and then split the changes in https://github.com/errordeveloper/kubernetes/pull/17 up into separate PRs against kubernetes master so they can be discussed and merged separately.

We discussed the `etcd` flags change in #sig-cluster-lifecycle on Slack and @pires (the author of the external etcd code) said:
> The flags exist for multiple different setups. We can have client authentication without TLS, etc. My pr described multiple scenarios.

Therefore I think we should discuss those use-cases before making any changes to this. It was felt that this change doesn't need to block merging this PR.

lukemarsden commented:
Hey @simonswine I'm glad you're finding this useful! We do hope to include an alpha version of `kubeadm` in the 1.4 stable release channel so that it can be used to install stable 1.4 clusters.

lukemarsden commented:
Update: @luxas is making some changes in https://github.com/errordeveloper/kubernetes/pull/17 that we do want to land, so we'll test those and get them in tomorrow morning UK time.

luxas committed:
Various improvements plus added a version command

lukemarsden commented:
Summarizing changes in @luxas' latest changeset:
* Remove `--schedule-pods-here` in favour of documenting `kubectl taint` (addresses @mikedanese's review feedback)
* Friendlier `--cloud-provider` validation for users.
* `kubeadm version` command which just returns the Kubernetes version.
* Switch to official discovery image
* Don't expose cm and scheduler ports to entire world

lukemarsden commented:
Thanks @luxas and everyone else!

luxas commented:
And no other PRs in SQ! That's great

luxas committed:
Update CHANGELOG and gofmt

justinsb commented:
@k8s-bot gke e2e test this

lukemarsden committed:
fix https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/33262/kubernetes-pull-verify-all/15586/

mikedanese commented:
Please drop merge commits.

lukemarsden commented:
> Please drop merge commits.

Done.

errordeveloper commented:
Mike, do you have any poi gets to docs to confirm that?

On Tue, 27 Sep 2016, 11:53 Mike Danese, <notifications@github.com> wrote:

> *@mikedanese* commented on this pull request.
> ------------------------------
>
> In cmd/kubeadm/app/api/types.go
> <https://github.com/kubernetes/kubernetes/pull/33262>:
>
> > +		}
> +	}
> +	Services struct {
> +		CIDR      net.IPNet
> +		DNSDomain string
> +	}
> +	PodNetwork struct {
> +		CIDR net.IPNet
> +	}
> +	CloudProvider string
> +	Schedulable   bool
> +}
> +
> +const (
> +	DefaultServiceDNSDomain   = "cluster.local"
> +	DefaultServicesCIDRString = "100.64.0.0/12"
>
> Both AWS and Google networks use this range. This is going to cause weird
> failures.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/33262>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAPWS6gJKEwsWwyigzkctI6F1oMG74_uks5quPWxgaJpZM4KDq18>
> .
>
